arangod/Agency/TransactionBuilder.h

```diff
@@ -32,210 +32,214 @@
+      _builder->openObject();
+      _builder->add("oldEmpty", VPackValue(true));
+      _builder->close();
+      return std::move(*this);
+    }
+
+    envelope end() {
+      _builder->close();
+      _builder->add(VPackValue(AgencyWriteTransaction::randomClientId()));
+      return envelope(_builder.release());
+    }
+    ~precs_trx() try {
+      if (_builder) {
+        end();
+      }
+    } catch (...) {
+    }
+    precs_trx(precs_trx&&) = default;
+    precs_trx& operator=(precs_trx&&) = default;
+
+   private:
+    friend write_trx;
+    precs_trx(builder_ptr b) : _builder(std::move(b)) { _builder->openObject(); }
+    builder_ptr _builder;
+  };
+
+  struct write_trx {
+    envelope end() {
+      _builder->close();
+      _builder->add(VPackSlice::emptyObjectSlice());
+      _builder->add(VPackValue(AgencyWriteTransaction::randomClientId()));
+      return envelope(_builder.release());
+    }
+    template <typename K, typename V>
+    write_trx&& key(K&& k, V&& v) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      detail::add_to_builder(_builder.get(), std::forward<V>(v));
+      return std::move(*this);
+    }
+    template <typename K, typename F>
+    write_trx&& emplace(K&& k, F f) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      f(*_builder);
+      _builder->close();
+      return std::move(*this);
+    }
+
+    template <typename K, typename V>
+    write_trx&& set(K&& k, V&& v) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      _builder->add("op", VPackValue("set"));
+      this->set("new", std::forward<V>(v));
+      _builder->close();
+      return std::move(*this);
+    }
+
+    template <typename K>
+    write_trx&& remove(K&& k) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      _builder->add("op", VPackValue("delete"));
+      _builder->close();
+      return std::move(*this);
+    }
+
+    template <typename K>
+    write_trx&& inc(K&& k, uint64_t delta = 1) && {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      this->key("op", "increment");
+      this->key("delta", delta);
+      _builder->close();
+      return std::move(*this);
+    }
+
+    ~write_trx() try {
+      if (_builder) {
+        end();
+      }
+    } catch (...) {
+    }
+    write_trx(write_trx&&) = default;
+    write_trx& operator=(write_trx&&) = default;
+
+
+    precs_trx precs() { return precs_trx(std::move(_builder)); }
+
+   private:
+    friend envelope;
+    write_trx(builder_ptr b) : _builder(std::move(b)) {
+      _builder->openObject();
+    }
+    builder_ptr _builder;
+  };
+
+  read_trx read() { return read_trx(std::move(_builder)); }
+  write_trx write() { return write_trx(std::move(_builder)); }
+
+  void done() {
+    _builder->close();
+    _builder.release();
```

Comment:
Is the `_builder.release()` call intentional here? This will mean the value constructed in the builder is gone and useless.
If this method is only supposed to be called from the destructor, it is fine, but then it should be made private. If it is a public method, I don't see its use case if it includes the call to `_builder.release().